cpp/src/arrow/dataset/partition.cc

```diff
@@ -361,7 +411,32 @@ Result<std::string> DirectoryPartitioning::FormatValues(
     break;
   }
 
-  return fs::internal::JoinAbstractPath(std::move(segments));
+  return std::make_pair(fs::internal::JoinAbstractPath(std::move(segments)), "");
+}
+
+Result<std::pair<std::string, std::string>> FilenamePartitioning::FormatValues(
+    const ScalarVector& values) const {
+  std::vector<std::string> segments(static_cast<size_t>(schema_->num_fields()));
+
+  for (int i = 0; i < schema_->num_fields(); ++i) {
+    if (values[i] != nullptr && values[i]->is_valid) {
+      segments[i] = values[i]->ToString();
+      continue;
+    }
+
+    if (auto illegal_index = NextValid(values, i)) {
+      // XXX maybe we should just ignore keys provided after the first absent one?
+      return Status::Invalid("No partition key for ", schema_->field(i)->name(),
+                             " but a key was provided subsequently for ",
+                             schema_->field(*illegal_index)->name(), ".");
+    }
+
+    // if all subsequent keys are absent we'll just print the available keys
+    break;
```

Comment:
Are there tests in place for this case?  In other words, do we have a test where the schema is something like `schema({field("a", int32()), field("b", int32())}` and we only pass in one field (or no fields)?

Reply:
Added a test case with only one field, do we need any more changes?

Reply:
That should be fine for this case.